File: src/ros2_bridge/mappers/geometry_msgs/quaternion_stamped.rs
//! Typed mapper for `geometry_msgs/msg/QuaternionStamped`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn quaternion_stamped_to_bus(msg: ros_env::geometry_msgs::msg::QuaternionStamped) -> crate::geometry_msgs::msg::v1::QuaternionStamped {
    crate::geometry_msgs::msg::v1::QuaternionStamped {
        header: Some(crate::ros2_bridge::mappers::std_msgs::header::header_to_bus(msg.header)),
        quaternion: Some(crate::ros2_bridge::mappers::geometry_msgs::quaternion::quaternion_to_bus(msg.quaternion)),
    }
}

pub(crate) fn quaternion_stamped_to_ros(bus: crate::geometry_msgs::msg::v1::QuaternionStamped) -> ros_env::geometry_msgs::msg::QuaternionStamped {
    ros_env::geometry_msgs::msg::QuaternionStamped {
        header: crate::ros2_bridge::mappers::std_msgs::header::header_to_ros(bus.header.unwrap_or_default()),
        quaternion: crate::ros2_bridge::mappers::geometry_msgs::quaternion::quaternion_to_ros(bus.quaternion.unwrap_or_default()),
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct GeometryMsgsQuaternionStampedMapper;

impl TypedTopicMapper for GeometryMsgsQuaternionStampedMapper {
    type Ros = ros_env::geometry_msgs::msg::QuaternionStamped;
    type Bus = crate::geometry_msgs::msg::v1::QuaternionStamped;

    fn type_name(&self) -> &'static str {
        "geometry_msgs/msg/QuaternionStamped"
    }

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(quaternion_stamped_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(quaternion_stamped_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/geometry_msgs/quaternion_stamped.py
"""Generated mapper for `geometry_msgs/msg/QuaternionStamped`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert
from robot_bus.ros2_bridge.mappers.std_msgs.header import header_to_bus, header_to_ros
from robot_bus.ros2_bridge.mappers.geometry_msgs.quaternion import quaternion_to_bus, quaternion_to_ros

def quaternion_stamped_to_bus(msg):
    from robot_bus.geometry_msgs.msg.v1 import QuaternionStamped as BusMsg

    bus = BusMsg()
    bus.header.CopyFrom(header_to_bus(msg.header))
    bus.quaternion.CopyFrom(quaternion_to_bus(msg.quaternion))
    return bus


def quaternion_stamped_to_ros(bus):
    from geometry_msgs.msg import QuaternionStamped as RosMsg

    out = RosMsg()
    out.header = header_to_ros(bus.header)
    out.quaternion = quaternion_to_ros(bus.quaternion)
    return out


class GeometryMsgsQuaternionStampedMapper:
    def type_name(self) -> str:
        return "geometry_msgs/msg/QuaternionStamped"

    def ros_msg_type(self):
        from geometry_msgs.msg import QuaternionStamped as RosMsg

        return RosMsg

    def ros_to_bus(self, msg) -> bytes:
        return quaternion_stamped_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        from robot_bus.geometry_msgs.msg.v1 import QuaternionStamped as BusMsg

        bus = BusMsg()
        bus.ParseFromString(payload)
        return quaternion_stamped_to_ros(bus)
